spec/unit/coinbase/address_spec.rb

```diff
@@ -449,6 +452,238 @@
+          let(:normalized_from_asset_id) { 'eth' }
+          let(:amount) { 500_000_000 }
+
+          it 'returns the broadcast trade' do
+            expect(trade).to eq(broadcasted_trade)
+          end
+
+          it 'signs the transaction with the address key' do
+            expect(transaction).to have_received(:sign).with(key)
+          end
+        end
+
+        context 'when the asset is ETH' do
+          let(:from_asset_id) { :eth }
+          let(:amount) { 0.5 }
+          let(:trade_amount) { 500_000_000_000_000_000 }
+
+          it 'returns the broadcast trade' do
+            expect(trade).to eq(broadcasted_trade)
+          end
+
+          it 'signs the transaction with the address key' do
+            expect(transaction).to have_received(:sign).with(key)
+          end
+        end
+
+        context 'when the asset is USDC' do
+          let(:from_asset_id) { :usdc }
+          let(:normalized_from_asset_id) { 'usdc' }
+          let(:amount) { 5 }
+          let(:trade_amount) { 5_000_000 }
+          let(:balance_response) { usdc_balance_response }
+
+          it 'creates a Trade' do
+            expect(trade).to eq(broadcasted_trade)
+          end
+        end
+
+        context 'when there is an approve transaction' do
+          let(:approve_signed_payload) { 'approve_signed_payload' }
+          let(:approve_transaction) { double(Coinbase::Transaction, sign: approve_signed_payload) }
+
+          let(:broadcast_trade_request) do
+            {
+              signed_payload: signed_payload,
+              approve_transaction_signed_payload: approve_signed_payload
+            }
+          end
+
+          it 'creates a Trade' do
+            expect(trade).to eq(broadcasted_trade)
+          end
+
+          it 'signs the trade transaction with the address key' do
+            expect(transaction).to have_received(:sign).with(key)
+          end
+
+          it 'signs the approve transaction with the address key' do
+            expect(approve_transaction).to have_received(:sign).with(key)
+          end
+        end
+      end
+    end
+
+    describe 'when the address cannot sign' do
+      let(:unhydrated_address) { described_class.new(model, nil) }
+
+      it 'raises an error' do
+        expect do
+          unhydrated_address.trade(12_345, from_asset_id, to_asset_id)
+        end.to raise_error('Cannot trade from address without private key loaded')
+      end
+    end
+
+    describe 'when the from asset is unsupported' do
+      it 'raises an ArgumentError' do
+        expect do
+          address.trade(amount, :uni, to_asset_id)
+        end.to raise_error(ArgumentError, 'Unsupported from asset: uni')
+      end
+    end
+
+    describe 'when the to asset is unsupported' do
+      it 'raises an ArgumentError' do
+        expect do
+          address.trade(amount, from_asset_id, :uni)
+        end.to raise_error(ArgumentError, 'Unsupported to asset: uni')
+      end
+    end
```

Comment:
duplicated test

Reply:
@erdimaden which are the duplicated tests?

Reply:
My bad, I misread that.